File: 2670. Find the Distinct Difference Array/solution.py
class Solution(object):
  def distinctDifferenceArray(self, nums):
    """
    :type nums: List[int]
    :rtype: List[int]
    """
    prefix = set()
    suffix = {}
    for num in nums:
      if num not in suffix:
        suffix[num] = 1
      else:
        suffix[num] += 1
    
    result = []

    for num in nums:
      prefix.add(num)
      
      suffix[num] -= 1
      if suffix[num] == 0:
        suffix.pop(num)
      
      result.append(len(prefix) - len(suffix))

    return result

File: 2670. Find the Distinct Difference Array/solution.cpp
class Solution {
public:
  vector<int> distinctDifferenceArray(vector<int>& nums) {
    unordered_set<int> prefix;
    unordered_map<int, int> suffix;
    for (int index = 0; index < nums.size(); index++){
      int num = nums[index];
      suffix[num]++;
    }
    
    vector<int> result;

    for (int index = 0; index < nums.size(); index++){
      int num = nums[index];
      prefix.insert(num);
      
      suffix[num] -= 1;
      if (suffix[num] == 0){
        suffix.erase(num);
      }
      
      result.push_back(prefix.size() - suffix.size());
    }

    return result;
  }
};